components/brave_wallet/browser/unstoppable_domains_multichain_calls.h

```diff
@@ -0,0 +1,79 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_UNSTOPPABLE_DOMAINS_MULTICHAIN_CALLS_H_
+#define BRAVE_COMPONENTS_BRAVE_WALLET_BROWSER_UNSTOPPABLE_DOMAINS_MULTICHAIN_CALLS_H_
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "base/callback.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_wallet {
+
+template <class ResultType>
+class UnstoppableDomainsMultichainCall {
+ public:
+  using CallbackType = base::OnceCallback<
+      void(const ResultType&, mojom::ProviderError, const std::string&)>;
+
+  struct Response {
+    absl::optional<ResultType> result;
+    absl::optional<mojom::ProviderError> error;
+    absl::optional<std::string> error_message;
+  };
+
+  UnstoppableDomainsMultichainCall() = default;
+  ~UnstoppableDomainsMultichainCall() = default;
+
+  bool SetNoResult(const std::string& chain_id);
+  bool SetResult(const std::string& chain_id, ResultType result);
+  bool SetError(const std::string& chain_id,
+                mojom::ProviderError error,
+                std::string error_message);
+  void AddCallback(CallbackType cb);
+
+ private:
+  Response* GetEffectiveResponse();
+  bool MaybeResolveCallbacks();
+
+  // chain_id -> response.
+  std::map<std::string, Response> responses_;
```

Comment:
That's my personal preference:)
I believe it would be hard to find measurable perf difference between these two for UD case, but changed to flat_map as it looks like it is much more frequent in code around.